### PR TITLE
Defect/PM-1196 - SSO with Email 2FA Flow - Email Required error fixed

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## Type of change

```
- [X] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

[PM-1196](https://bitwarden.atlassian.net/browse/PM-1196)
To fix the non-functional SSO with email 2FA flow which currently breaks with an "email required" error. 

The fix required sending the email from the IDP from the server to the client and setting it on the auth service so that the 2FA component could access it as part of the send email request.  Further, we also had to create a user specific session token to serve as an authentication factor for the anonymous send email endpoint to avoid malicious use of the endpoint (as we didn't have access to the master password hash or any other verifiable secret) 

See client PR here: [#5280](https://github.com/bitwarden/clients/pull/5280)

## Code changes

* **src/Core/Auth/Models/Business/Tokenables/SsoEmail2faSessionTokenable.cs:** Create a tokenable for the SSO Login with Email 2FA case
* **src/SharedWeb/Utilities/ServiceCollectionExtensions.cs:** Add a singleton service token factory for the new `SsoEmail2faSessionTokenable`
* **src/Identity/IdentityServer/BaseRequestValidator.cs:** If we have email 2FA required, send down user email and the SsoEmail2faSessionToken as part of the Two factor rejection. 
* **src/Api/Auth/Controllers/TwoFactorController.cs:** Update `SendEmailLoginAsync` to handle verifying a `SsoEmail2FaSessionToken` and sending a 2FA email. 

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

[PM-1196]: https://bitwarden.atlassian.net/browse/PM-1196?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Api/Auth/Controllers/TwoFactorController.cs [hunk 1]
```diff
@@ -285,30 +295,46 @@ public async Task SendEmail([FromBody] TwoFactorEmailRequestModel model)
...
+            else if (await _userService.VerifySecretAsync(user, requestModel.Secret))
             {
                 await _userService.SendTwoFactorEmailAsync(user);
                 return;
             }
         }
 
-        await Task.Delay(2000);
-        throw new BadRequestException("Cannot send two-factor email.");
+        this.ThrowDelayedBadRequestExceptionAsync(
```
Comment (justindbaur) [inline src/Api/Auth/Controllers/TwoFactorController.cs:336]: Missing an `await`?

File: src/Api/Auth/Models/Request/Accounts/SecretVerificationRequestModel.cs [hunk 2]
```diff
@@ -14,7 +14,7 @@ public virtual IEnumerable<ValidationResult> Validate(ValidationContext validati
     {
         if (string.IsNullOrEmpty(Secret) && string.IsNullOrEmpty(AuthRequestAccessCode))
         {
-            yield return new ValidationResult("MasterPasswordHash, OTP or AccessCode must be supplied.");
+            yield return new ValidationResult("MasterPasswordHash, OTP, or AccessCode must be supplied.");
```
Comment (justindbaur) [inline src/Api/Auth/Models/Request/Accounts/SecretVerificationRequestModel.cs:17]: Oxford comma 😍

File: src/Api/Auth/Models/Request/TwoFactorRequestModels.cs [hunk 3]
```diff
@@ -225,6 +225,14 @@ public User ToUser(User extistingUser)
...
         return extistingUser;
     }
+
+    public override IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
+    {
+        if (string.IsNullOrEmpty(Secret) && string.IsNullOrEmpty(AuthRequestAccessCode) && string.IsNullOrEmpty((SsoEmail2FaSessionToken)))
+        {
+            yield return new ValidationResult("MasterPasswordHash, OTP, AccessCode, or SsoEmail2faSessionToken must be supplied.");
+        }
+    }
```
Comment (justindbaur) [inline src/Api/Auth/Models/Request/TwoFactorRequestModels.cs:235]: Should this call into the base `Validate` method? I'm not sure if it actually does anything or not.

File: src/Identity/IdentityServer/CustomTokenRequestValidator.cs [hunk 4]
```diff
@@ -37,11 +39,11 @@ public class CustomTokenRequestValidator : BaseRequestValidator<CustomTokenReque
         IPolicyRepository policyRepository,
         ISsoConfigRepository ssoConfigRepository,
         IUserRepository userRepository,
-        IPolicyService policyService)
+        IPolicyService policyService, IDataProtectorTokenFactory<SsoEmail2faSessionTokenable> tokenDataFactory)
```
Comment (justindbaur) [inline src/Identity/IdentityServer/CustomTokenRequestValidator.cs:42]: nit: Put this on a new line.

**Review by justindbaur — CHANGES_REQUESTED** [review]
Just a few things, looks great though.

File: src/Api/Auth/Controllers/TwoFactorController.cs (hunk 1 above)
Comment (JaredSnider-Bitwarden) [reply src/Api/Auth/Controllers/TwoFactorController.cs:336]: image

I thought so too initially, but it's not awaitable b/c it doesn't return anything and just throws.

Comment (JaredSnider-Bitwarden) [reply src/Api/Auth/Controllers/TwoFactorController.cs]: Ah, thank you for the explanation. I did not know that. I've updated it to be `async Task` after some research.

**Review by justindbaur — APPROVED** [review]
Fantastic!

**Comment by JaredSnider-Bitwarden** [comment]
Passed QA; merging!

**Pull request merged**